examples/nlp/entity_linking/build_and_query_index.py

```diff
@@ -0,0 +1,290 @@
+
+    emb_file = h5py.File(cfg.embedding_save_name, "w")
+    emb_file.create_dataset(cfg.index_ds.name, data=embeddings)
+    emb_file.close()
+
+    pkl.dump(concept_ids, open(cfg.concept_id_save_name, "wb"))
+
+    return embeddings, concept_ids
+
+
+def get_query_embedding(query, model):
+    """Use entity linking encoder to get embedding for index query"""
+    model_input = model.tokenizer(
+        query,
+        add_special_tokens=True,
+        padding=True,
+        truncation=True,
+        max_length=512,
+        return_token_type_ids=True,
+        return_attention_mask=True,
+    )
+
+    query_emb = model.forward(
+        input_ids=torch.LongTensor([model_input["input_ids"]]).to(device),
+        token_type_ids=torch.LongTensor([model_input["token_type_ids"]]).to(device),
+        attention_mask=torch.LongTensor([model_input["attention_mask"]]).to(device),
+    )
+
+    return query_emb
+
+
+def query_index(
+    query: str, top_n: int, cfg: DictConfig, model: object, index: object, pca: object, idx2id: dict, id2string: dict,
+) -> Dict:
+
+    """
+    Query the nearest neighbor index of entities to find the 
+    concepts in the index dataset that are most similar to the 
+    query.
+
+    Args:
+        query (str): entity to look up in the index
+        cfg (DictConfig): config object to specifiy query parameters
+        model (EntityLinkingModel): entity linking encoder model
+        index: faiss index
+        idx2id (dict): dictionary mapping unique concept dataset index to 
+                       its CUI
+        id2string (dict): dictionary mapping each unqiue CUI to a 
+                          representative english description of
+                          the concept
+    Returns:
+        A dictionary with the concept ids of the index's most similar 
+        entities as the keys and a tuple containing the string 
+        representation of that concept and its cosine similarity to 
+        the query as the values. 
+    """
+    query_emb = get_query_embedding(query, model).detach().cpu().numpy()
+
+    if cfg.apply_pca:
+        query_emb = pca.transform(query_emb)
+
+    dist, neighbors = index.search(query_emb.astype(np.float32), cfg.query_num_factor * top_n)
+    dist, neighbors = dist[0], neighbors[0]
+    unique_ids = OrderedDict()
+    neighbor_idx = 0
+
+    # Many of nearest neighbors could map to the same concept id, their idx is their unique identifier
+    while len(unique_ids) < top_n and neighbor_idx < len(neighbors):
+        concept_id_idx = neighbors[neighbor_idx]
+        concept_id = idx2id[concept_id_idx]
+
+        # Only want one instance of each unique concept
+        if concept_id not in unique_ids:
+            concept = id2string[concept_id]
+            unique_ids[concept_id] = (concept, 1 - dist[neighbor_idx])
+
+        neighbor_idx += 1
+
+    unique_ids = dict(unique_ids)
+
+    return unique_ids
+
+
+def main(cfg: DictConfig, top_n: int, restore: bool):
```

Comment:
Shouldn't top_n belong in either the config file or in the query itself? Feels like a config param. Scope-wise it's more relevant to the query than the index building and could change from one client to the next.

Reply:
Oh I see. It's really "load and query" and if it's not built, go build it first. I think this should be separated into a build stage and a query utility, where the query utility gives a constructive error if not already built. Best to avoid the "surprise! I'm going to go build now."